[홈 탭] 홉 탭 디자인 및 기능 구현

s9hn opened the pull request:
## 😋 작업한 내용

<!-- 아래 리스트를 지우고, 작업 내용을 적어주세요. -->
- 홈 탭 디자인 구성
- 홈 탭 카드 뷰페이저 구현
  - 뷰페이저 커스텀 구현
- 홈 탭 잔디 리사이클러뷰 구현
- 홈 탭 선택 투두 리사이클러뷰 구현

## 🙏 PR Point

<!-- 피드백을 받고 싶은 부분이나, 공유하고 싶은 부분을 적어주세요. -->

### 리사이클러뷰 최적화
리사이클러뷰 최적화를 위해 적용한 방법은 아래와 같습니다.
1. RecyclerView.Adapter -> ListAdapter 전환
- 리스트 어댑터는 리사이클러뷰 어댑터가 지원하는 모든 기능을 지원합니다.
- DiffUtil 구현을 쉽게 합니다.
  - DiffUtil은 아이템 변경에 대한 notify를 명시하지 않고도 아이템의 변경 사항을 전달할 수 있습니다.
  - 아이템 변경 적용을 메인스레드가 아닌 백그라운드 스레드에서 진행합니다.

2. setHasFixedSize(true)
- 리사이클러뷰는 아이템이 변경될 때마다, 사이즈가 변경될 수 있는 View이며, 초기화마다 onMeasure, onDraw와 같은 메서드를 호출합니다.
- 해당 메서드를 true로 바꾸어주어, size가 고정임을 명시하고, 비용이 큰 메서드 호출을 막습니다.

3.  setHasStableIds(true)
- 아이템마다 포지션을 고유ID로 갖는것이 아닌, 아이템 자체의 고유ID를 이용해 아이템마다 고유ID를 부여합니다.
- 아이템마다 고유ID를 가지고 있으므로, 똑같은 아이템을 띄울경우, onBindViewHolder를 호출하지 않고, 이미 만들어진 해당 아이템을 재사용합니다.

- ~~리스트 어댑터를 사용하고, 고유 ID를 부여하였음에도 선택투두 아이템의 깜빡임 현상이 해결되지 않습니다. 아시는분 댓 ㅎ~~
- 커스텀 뷰페이저에 대한 코드는.. 긁어온거라 완벽히 이해하지는 않았습니다 이것저것 draw하는것만 이해함스..


## 👍 관련 이슈

<!-- 작업한 이슈번호를 # 뒤에 붙여주세요. -->

- Resolved : #7

s9hn committed:
feat: homeFragment 뷰 디자인 완성

1. viewPager2 의존성 추가
2. selector 추가
3. 리사이클러뷰, 뷰페이저에 들어갈 아이템 뷰 객체 추가
4. 뷰에 적용될 배경 selector 및 rectangle 리소스 추가
5. 스트링 적용

s9hn committed:
refactor: Lint 체크!

s9hn committed:
feat: 홈 뷰 디자인은 완성

s9hn committed:
feat: 뷰페이저 기능 완성

s9hn committed:
feat: 스터디카드 뷰페이저 어댑터 데이터바인딩

s9hn committed:
feat: 뷰페이저2를 이용한 홈 뷰 카드 구현

s9hn committed:
refactor: 잔디 리사이클러뷰 제거 및 선택투두 리사이클러뷰 연결

s9hn committed:
feat: 잔디 리사이클러뷰 어댑터 추가

s9hn committed:
feat: 홈뷰 Dto 매핑 도메인 모델 생성

s9hn committed:
feat: 홈 뷰모델 생성 및 더미데이터 생성

s9hn committed:
feat: 더미데이터, 어댑터들 연결

1. 필요한 데이터 ui모델 생성

s9hn committed:
feat: 스터디원 투두 진행율 잔디갯수로 변환하는 로직 추가

s9hn committed:
refactor: GrassUiModel 수정

s9hn committed:
feat: 잔디 어댑터 구현 및 바인딩

s9hn committed:
feat: 잔디 selected 바인딩어댑터 유틸함수 추가

s9hn committed:
feat: 잔디 회전 및 버그 수정

s9hn committed:
refactor: 커스텀 뷰페이저 리팩터링

s9hn committed:
feat: 투두 및 카드에 클릭리스너 생성 및 세팅

s9hn committed:
feat: 투두 체크업데이트 기능 끝~

s9hn committed:
refactor: 주석 수정 및 로그 제거

s9hn committed:
refactor: homeUiModel -> StudyUiModel 이름 변경

s9hn committed:
test: homeUiModel 잔디갯수 출력메서드 테스트

s9hn committed:
refactor: StudyUiModel에 studyId 프로퍼티 추가

s9hn committed:
refactor: 리사이클러뷰 성능개선

1. DiffUtil을 사용한 리스트어댑터 구현
2. setHasStableIds(true)로 아이템마다 고유 ID 부여 -> onBindViewHolder 호출 감소

s9hn committed:
refactor: 더미데이터 수정

s9hn committed:
refactor: 린트정리

s9hn committed:
fix: 투두 깜빡임 문제 해결

RightHennessy commented on android/app/src/main/java/com/created/team201/presentation/home/HomeFragment.kt:
```diff
@@ -2,13 +2,55 @@ package com.created.team201.presentation.home
+
+    private fun initViewModel() {
+        binding.viewModel = homeViewModel
+        binding.lifecycleOwner = this
```
여기서 this는 어떤 lifecycle인가요?
observe 하실 때 사용하신 viewLifeCycleOwner와 어떤 차이점이 있어 다르게 사용하셨는지 궁금합니다.

s9hn replied:
this는 fragment의 lifecycle입니다. 
어떤 라이프사이클을 참조하더라도, onDestroyView에서 바인딩 참조를 해제하고 있으므로, 메모리 누수 등이 일어날 일은 없습니다.

binding.lifecycleOwner 는 바인딩 된 뷰모델이 어떤 생명주기를 따를 것인지 명시해주는 역할입니다.
해당 코드가 있어야만, 바인딩된 xml이 바인딩된 뷰모델의 라이브데이터를 따라 실시간으로 ui를 갱신할 수 있습니다.

```kotlin
  homeViewModel.observe(lifecycleOwner) {   }
```
해당 코드 2줄은 위 코드와 유사합니다.
fragment에서는 fragment의 자체 라이프사이클을 따르는 것은 위험하므로, viewLifeCycleOwner를 할당해주었습니다.

RightHennessy commented on android/app/src/main/java/com/created/team201/presentation/home/HomeFragment.kt:
```diff
@@ -2,13 +2,55 @@ package com.created.team201.presentation.home
+    }
+
+    private fun initAdapter() {
+        customViewPager.setCustomViewPager()
```
CustomViewPager 내부에서 init 블럭을 통해 초기화 하지 않고 setCustoViewPager를 직접 호출해주시는 이유가 있을까요?

s9hn replied:
뷰페이저의 세팅을 단순히 객체를 생성할 때 호출되는 트리거가 아닌, 외부에서 해당 객체의 함수를 트리거로써 호출하는게 옳다고 생각했습니다.

s9hn replied:
가독성 측면에서 해당 코드로 세팅해주었슴다

RightHennessy replied:
납득했습니다.

RightHennessy commented on android/app/src/main/java/com/created/team201/presentation/home/HomeViewModel.kt:
```diff
@@ -0,0 +1,108 @@
+
+    private fun updateOptionalTodoCheck(studies: List<StudyUiModel>, id: Int, isDone: Boolean) {
+        _userStudies.value = studies.map { studyUiModel ->
+            studyUiModel.takeIf { todoUiModel -> !todoUiModel.optionalTodos.any { it.todoId == id } }
```
takeIf 함수.. 매우 흥미롭군요 🧐

RightHennessy commented on android/app/src/main/java/com/created/team201/presentation/home/adapter/viewholder/DashboardViewHolder.kt:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    companion object {
+        fun getView(parent: ViewGroup, layoutInflater: LayoutInflater): ItemHomeBinding =
+            ItemHomeBinding.inflate(layoutInflater, parent, false)
+    }
```
오.. binding을 이렇게 전달할 수도 있군요.. 하나 배워갑니다.

no1msh replied:
산군에 대한 리뷰는 아니지만 ㅎㅎ 메시지를 보내는 구조로 만들기위해 저렇게 구성하는거로 알고있습니다.

RightHennessy commented on android/app/src/main/java/com/created/team201/presentation/home/adapter/viewholder/DashboardViewHolder.kt:
```diff
@@ -0,0 +1,48 @@
+        attachHomeGrassAdapter()
+    }
+
+    private fun attachOptionalTodoAdapter() {
+        binding.rvHomeOptionalTodoList.setHasFixedSize(true)
+        binding.rvHomeOptionalTodoList.itemAnimator = null
+        binding.rvHomeOptionalTodoList.adapter = optionalTodoAdapter
+    }
```
apply로 코드를 묶으면 가독성이 더 높아질 수 있지 않을까여?

s9hn replied:
뭐가 달라지나요? 저도 원래 with(binding) 으로 다 묶었는데,,
뎁스 하나 더 깊어지고, 별 다른게 없어서,, apply써도 3줄아닌가요 ..

no1msh replied:
저도 이건 취향차이라고 보는데 개인적으론 뎁스 들어가는게 싫어서 저도 안씁니다.

RightHennessy commented on android/app/src/main/res/layout/item_home.xml:
```diff
@@ -0,0 +1,165 @@
+            android:layout_height="wrap_content"
+            android:layout_marginStart="16dp"
+            android:layout_marginTop="16dp"
+            android:text="선택투두"
```
string으로 ㅃ ㅐ줘잉

RightHennessy commented on android/app/src/main/res/layout/item_home_grass.xml:
```diff
@@ -0,0 +1,33 @@
+            android:id="@+id/iv_home_grass_blank"
+            android:layout_width="25dp"
+            android:layout_height="25dp"
+            android:layout_marginStart="2dp"
+            android:layout_marginTop="2dp"
+            android:layout_marginEnd="2dp"
+            android:layout_marginBottom="2dp"
```
?~! 왜죠?

s9hn replied:
피그마 디자이너가 간격을 저렇게 줬습니다.

RightHennessy commented on android/app/src/main/java/com/created/team201/presentation/home/adapter/DashboardAdapter.kt:
```diff
@@ -0,0 +1,45 @@
+        setHasStableIds(true)
+    }
+
+    override fun getItemId(position: Int): Long {
+        return getItem(position).studyId.toLong()
+    }
```
id를 처음부터 Long으로 저장하면 변환작업이 필요없지 않을까요?

RightHennessy reviewed (commented):
さんくん ~ 너무 수고햇졍..
특히나 잔디 UPDATE 하는 로직은 .. 매우 복잡하구만

s9hn committed:
refactor: 잔디뷰 마진 수정

s9hn committed:
refactor: uiModel, domainModel 프로퍼티 타입 Int -> Long 수정

s9hn committed:
refactor: 홈뷰 문자열 스트링추출

no1msh commented on android/app/src/main/res/layout/item_home.xml:
```diff
@@ -0,0 +1,165 @@
+            android:textColor="@color/grey02_78808B"
+            app:layout_constraintEnd_toEndOf="@+id/rv_home_grass"
+            app:layout_constraintStart_toStartOf="@+id/rv_home_grass"
+            app:layout_constraintTop_toBottomOf="@+id/rv_home_grass"
```
이미 `id`가 존재하는 뷰를 참조하는 것이니 `+`를 떼어 주시는게 좋을 듯하네요

no1msh commented on android/app/src/main/java/com/created/team201/presentation/home/CustomViewPager.kt:
```diff
@@ -0,0 +1,91 @@
+        binding.vpHome.offscreenPageLimit = 1
+    }
+
+    inner class ZoomOutPageTransformer : ViewPager2.PageTransformer {
+
+        override fun transformPage(view: View, position: Float) {
+            view.apply {
+                val pageWidth = width
+                val pageHeight = height
+
+                when {
+                    position < -1 -> initializeAlpha(0f)
+                    position <= 1 -> setScale(position, pageHeight, pageWidth)
+                    else -> initializeAlpha(0f)
+                }
+            }
+        }
+
+        private fun View.setScale(
+            position: Float,
+            pageHeight: Int,
+            pageWidth: Int,
+        ) {
+            val scaleFactor = max(MIN_SCALE, 1 - abs(position))
+            val verticalMargin = pageHeight * (1 - scaleFactor) / 2
+            val horizontalMargin = pageWidth * (1 - scaleFactor) / 2
+
+            translationX =
+                if (position < 0) horizontalMargin - verticalMargin / 2 else horizontalMargin + verticalMargin / 2
+            scaleX = scaleFactor
+            scaleY = scaleFactor
+
+            initializeAlpha((MIN_ALPHA + (((scaleFactor - MIN_SCALE) / (1 - MIN_SCALE)) * (1 - MIN_ALPHA))))
+        }
+
+        private fun View.initializeAlpha(f: Float) {
+            alpha = f
+        }
+    }
```
inner class로 구현하신 이유가 궁금합니다!

s9hn replied:
본 클래스에서만 사용될 유틸이기에 캡슐화를 위해 이너클래스로 두었습니다

s9hn committed:
refactor: BuildConfig 임포트

s9hn committed:
refactor: 리뷰반영

1. lifecycleOwner this -> viewLifecycleOwner
2. initViewModel함수명 수정
3. userStudy 옵저브함수 시그니처? 수정

no1msh reviewed (changes requested):
さんくん!!
야하리 코딩실력이 스바라시데스가, 에..몇가지 코멘토를 남겼으까라, 확인 부탁드린다데스

RightHennessy reviewed (approved):
리뷰 반영 확인 굿뜨 👍

s9hn committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2023-yigongil into AN/feature/7-home-inflate

s9hn committed:
refactor: 뷰 참조 id값 시그니처수정
